lib/map.js

```diff
@@ -32,12 +38,26 @@ function map(values, iterator, extensions, done) {
     return done(null, results);
   }
 
-  for (idx = 0; idx < length; idx++) {
-    var key = keys[idx];
-    next(key);
+  var maxConcurrent = getMaxConcurrentIterations();
+  var running = 0;
+  var sync = false;
```

Comment:
What is this? I don't really understand it.

Reply:
`now-and-later` allows you to pass callbacks that are entirely sync, which means while the kickoff loop is running, a callback can complete - without this flag we'd reenter an inner `kickoff` loop, which, as written now isn't so bad (since we'll continue from either place in the same fashion), but needlessly increases the stack depth - by monitoring when tasks are completing in a `sync` manner, we can keep the callstack depth for a list of `sync` callbacks minimal, while only async callbacks which are beyond the maximal concurrent task count will increase the depth.

Reply:
Interesting. This gets me thinking about some other stuff but that is outside the scope of this.